### PR TITLE
feat(go): CTS implementation for go client

### PR DESCRIPTION
**Pull request by mehmetaligok**
## 🧭 What and Why

🎟 JIRA Ticket: https://algolia.atlassian.net/browse/APIC-661

The CTS suite for go client implemented. 

### Changes included:

- Test templates created for the generation. 
    - When creating tests, json unmarshaller is used to parse requests. Initially, I tried to create the request the regular way but couldn't do it properly. The main issues were mostly about different "oneof" types and some very detailed requests. Unfortunately, generating these in Go with the mustache is not easy since mustache is all about "Logic-less" templates. 
    - Maybe we can revisit this later. We may need some improvements for CTS in general because creating CTS stuff is hard to manage and not straightforward. 
- `extras` field added to test json specs.  
    - It is added for `skipForGo` prop. It was needed for three specific test cases. The client was working okay but the test unmarshalling was not working properly because of the payload. No need to block this improvement for this.
    - These `extras` can be also used for some other cases in the future because the only request may not be enough for all the language generations. 
- Added some improvements for client generation. 
    - `UnmarshalJSON` implemented for request struct to handle json parsing for the direct request. 
    - "OneOf" model unmarshaling updated to add an extra validation.

## 🧪 Test

The test should run with CI.

### CODE REVIEW

File: scripts/cts/runCts.ts [hunk 1]
```diff
@@ -22,6 +22,17 @@ async function runCtsOne(language: string): Promise<void> {
       );
       break;
     }
+    case 'go':
+      if (DOCKER) {
```
[the remark this answers is not in the document]
Comment (mehmetaligok) [reply scripts/cts/runCts.ts:26]: Unfortunately, it is in the `PATH` but somehow it is not working with js runtime. I can run with just `go` from the container but when I used it with the [run](https://github.com/algolia/api-clients-automation/blob/main/scripts/common.ts#L78) function it is not working. 

I couldn't quite figure out the reason last time. So I already opened a separate ticket to resolve this if possible.

File: tests/CTS/methods/requests/predict/updateSegment.json [hunk 2]
```diff
@@ -55,6 +55,9 @@
           ]
         }
       }
+    },
+    "extras": {
+      "skipForGo": true
```
[the remark this answers is not in the document]
Comment (mehmetaligok) [reply tests/CTS/methods/requests/predict/updateSegment.json:60]: Yes mainly because of "oneof". The contents of "oneof"s are very similar for this request. Since I used the json unmarshalling, the expected request body is not matching completely. It is unmarshalling it to a different "oneof" because it is matching with more than one statement. 

When I tried to create the request with a constructor it worked without an issue, I couldn't make it work for the tests.  I have opened another ticket to improve some of the CTS workflow for development. Maybe that can ease the development so we can take a look at this later on.

**Comment by mehmetaligok** [comment]
> Nice ! But it looks like the tests are not generated on the CI, I think it's an artifact issue, can you try to tweak `.github/actions/restore-artifacts/action.yml` file to include cts ?

@millotp Oh, I tried on local but I haven't noticed the test code does not exist on the generated branch. It is my bad. Let me take a look and update here. Thanks for the heads up.

**Comment by mehmetaligok** [comment]
@millotp push problem for the generated branch should be resolved. It is ready to be reviewed again. I also do some extra tests for negative and positive CI scenarios to ensure the `cts run` is working properly for the CI. I believe everything is in place right now.

File: templates/go/tests/requests/requests.mustache [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    Region:    {{clientPrefix}}.{{#lambda.uppercase}}{{defaultRegion}}{{/lambda.uppercase}},{{/hasRegionalHost}}
+    Requester: echo,
+  }
+  client := {{clientPrefix}}.NewClientWithConfig(cfg)
+
+  return client, echo
+}
+
+{{#blocksRequests}}
+func Test{{#lambda.titlecase}}{{clientPrefix}}{{/lambda.titlecase}}_{{#lambda.titlecase}}{{operationId}}{{/lambda.titlecase}}(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (mehmetaligok) [reply templates/go/tests/requests/requests.mustache:28]: since currently, we are using only an echo requester adding `t.Parallel()` is bringing flakiness to the tests. Some tests start randomly failing. 

We may need to add some complexity to test cases to handle this, I guess we can skip this at this phase.

[the remark this answers is not in the document]
Comment (millotp) [reply templates/go/tests/requests/requests.mustache:28]: no worries the test runs very fast already

**Review by aallam — APPROVED** [review]

**Comment by aallam** [comment]
Nice work! and as you mentioned, I am for revisiting the CTS specs for easier tests generation! Well done! 🚀

File: templates/go/tests/requests/requests.mustache [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        headers := map[string]string{}
+        require.NoError(t, json.Unmarshal([]byte(`{{{.}}}`), &headers))
+        for k, v := range headers {
+          require.Equal(t, v, echo.header.Get(k))
+        }
+      {{/request.headers}}
+      {{#request.queryParameters}}
+        queryParams := map[string]string{}
+        require.NoError(t, json.Unmarshal([]byte(`{{{.}}}`), &queryParams))
+        for k, v := range queryParams {
```
Comment (millotp) [inline templates/go/tests/requests/requests.mustache:89]: I don't remember what we do in other languages, but it might be worth to do a strict comparaison here, and check that the number of query params is the same in `queryParams` and in `echo.query`

**Review by millotp — APPROVED** [review]
looks amazing ! somehow kotlin shows up in the generation this time @aallam 🤔

File: templates/go/tests/requests/requests.mustache (hunk 4 above)
Comment (mehmetaligok) [reply templates/go/tests/requests/requests.mustache:89]: oh, I miss this before merging. But I noted this to check out.
